example_configs/kafka-2_0_0.yml

```diff
@@ -0,0 +1,89 @@
+lowercaseOutputName: true
+
+rules:
+- pattern : kafka.server<type=(.+), name=(.+), clientId=(.+), topic=(.+), partition=(.*)><>Value
+  name: kafka_server_$1_$2
+  type: GAUGE
+  labels:
+    clientId: "$3"
+    topic: "$4"
+    partition: "$5"
+- pattern : kafka.server<type=(.+), name=(.+), topic=(.+), partition=(.*)><>Value
+  name: kafka_server_$1_$2
+  type: GAUGE
+  labels:
+    topic: "$3"
+    partition: "$4"
+- pattern : kafka.server<type=(.+), name=(.+), clientId=(.+), brokerHost=(.+), brokerPort=(.+)><>Value
+  name: kafka_server_$1_$2
+  type: GAUGE
+  labels:
+    clientId: "$3"
+    broker: "$4:$5"
```

Comment:
Should metrics really be broken down by broker? That seems a bit over-specific to me. You care about how much you're communicating to the cluster overall, not where a partition lives.

Reply:
Interesting question!

One thing I would see in favor for breaking things down by broker is environments where there are differences in latency: Assume you're spreading your kafka cluster over multiple data centers, then it is likely that some brokers are "further away" than others. Generally I assume that the approach with Prometheus is to expose "raw data", and have either Prometheus recording rules or the visualization side (Grafana/...) responsible for making sense of that raw data. Aggregating already here would hide that information.

With that said though: How else could this be aggregated on the exporter side?

Reply:
The issue is that such data could be too high cardinality to start with, per broker per client thread could add up pretty fast.

Reply:
Okay so this is every broker times every kafka client - that's not at all sane cardinality. This set of metrics should be removed.

Reply:
You got me a bit worried here, so I did a check to see what kind of metrics we're actually getting with these rules. IT turns out that this "client id" is not a "user client of kafka", but rather refers to some internal clients related to the replication system:

~~~~
$ echo beans | java -jar ~/Downloads/jmxterm-1.0.0-uber.jar  -l 30371 | grep -E 'ConsumerLag'
Welcome to JMX terminal. Type "help" for available commands.
$>beans
#domain = JMImplementation:
#domain = com.sun.management:
#domain = java.lang:
#domain = java.nio:
#domain = java.util.logging:
#domain = kafka:
#domain = kafka.cluster:
#domain = kafka.controller:
#domain = kafka.coordinator.group:
#domain = kafka.coordinator.transaction:
#domain = kafka.log:
#domain = kafka.network:
#domain = kafka.server:
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=0,topic=collaborne-events,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=0,topic=ml1-errors,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=0,topic=ml1-events,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=0,topic=ml3-commands,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=0,topic=ml3-events,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=0,topic=ml3-query-responses,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=0,topic=ml4-errors,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=1,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=10,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=13,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=16,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=19,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=22,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=25,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=28,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=31,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=34,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=37,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=4,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=40,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=43,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=46,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=49,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-1,name=ConsumerLag,partition=7,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=0,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=0,topic=collaborne-commands,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=0,topic=collaborne-errors,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=0,topic=collaborne-query-responses,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=0,topic=meta-topic,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=0,topic=ml2-commands,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=0,topic=ml2-errors,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=0,topic=ml2-events,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=0,topic=ml3-errors,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=0,topic=qa-query-responses,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=12,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=15,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=18,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=21,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=24,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=27,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=3,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=30,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=33,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=36,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=39,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=42,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=45,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=48,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=6,topic=__consumer_offsets,type=FetcherLagMetrics
kafka.server:clientId=ReplicaFetcherThread-0-2,name=ConsumerLag,partition=9,topic=__consumer_offsets,type=FetcherLagMetrics
#domain = kafka.utils:
~~~~

~~~~
$ echo beans | java -jar ~/Downloads/jmxterm-1.0.0-uber.jar  -l 30371 | grep -E 'broker'
Welcome to JMX terminal. Type "help" for available commands.
$>beans
#domain = JMImplementation:
#domain = com.sun.management:
#domain = java.lang:
#domain = java.nio:
#domain = java.util.logging:
#domain = kafka:
#domain = kafka.cluster:
#domain = kafka.controller:
#domain = kafka.coordinator.group:
#domain = kafka.coordinator.transaction:
#domain = kafka.log:
#domain = kafka.network:
#domain = kafka.server:
#domain = kafka.utils:
kafka.server:broker-id=1,fetcher-id=0,type=replica-fetcher-metrics
$>kafka.server:broker-id=2,fetcher-id=0,type=replica-fetcher-metrics
kafka.server:brokerHost=localhost,brokerPort=9093,clientId=ReplicaFetcherThread-0-1,name=BytesPerSec,type=FetcherStats
kafka.server:brokerHost=localhost,brokerPort=9093,clientId=ReplicaFetcherThread-0-1,name=RequestsPerSec,type=FetcherStats
kafka.server:brokerHost=localhost,brokerPort=9094,clientId=ReplicaFetcherThread-0-2,name=BytesPerSec,type=FetcherStats
kafka.server:brokerHost=localhost,brokerPort=9094,clientId=ReplicaFetcherThread-0-2,name=RequestsPerSec,type=FetcherStats
~~~~

These metrics only appear when a topic is actually having multiple replicas, the above output is from a 1-zookeeper/3-kafka local test setup, with 10 "real" clients connected to about 30 topic/partitions.

> Beware: ConsumerLag is an overloaded term in Kafka—it can have the above definition, but is also used to represent the offset differences between partition leaders and their followers. If you see kafka.server:type=FetcherLagMetrics,name=ConsumerLag,clientId=([-.\w]+),topic=([-.\w]+),partition=([0-9]+), it is in reference to the latter, not the former.

(https://github.com/DataDog/the-monitor/blob/master/kafka/monitoring-kafka-performance-metrics.md)

> `kafka.server:type=FetcherLagMetrics,name=ConsumerLag,clientId=([-.\w]+),topic=([-.\w]+),partition=([0-9]+)`
> Lag in number of messages per follower replica. This is useful to know if the replica is slow or has stopped replicating from the leader.

(https://docs.confluent.io/current/kafka/monitoring.html)

Based on that I think it would be good to keep these metrics: There is a very valid use case here to see whether the replication works, and https://kafka.apache.org/documentation/#rep-throttle ("Ensuring Progress") explicitly refers to using the `ConsumerLag` metric.

Reply:
Okay, still a potential cardinality issue but a reasonable thing to have as you can't monitor replication otherwise.